src/ray.rs

```diff
@@ -315,6 +388,69 @@ mod tests {
         }
     }
 
+    /// Test whether a `Ray` which points at the center of a triangle
+    /// intersects it, unless it sees the back face, which is culled.
+    quickcheck!{
+        fn test_ray_hits_triangle(a: TupleVec,
+                                  b: TupleVec,
+                                  c: TupleVec,
+                                  origin: TupleVec,
+                                  u: u16,
+                                  v: u16)
+                                  -> bool {
+            // Define a triangle, u/v vectors and its normal
+            let triangle = (tuple_to_point(&a), tuple_to_point(&b), tuple_to_point(&c));
+            let u_vec = triangle.1 - triangle.0;
+            let v_vec = triangle.2 - triangle.0;
+            let normal = u_vec.cross(&v_vec);
+
+            // Get some u and v coordinates such that u+v <= 1
+            let u = u % 101;
+            let v = cmp::min(100 - u, v % 101);
+            let u = u as f32 / 100.0;
+            let v = v as f32 / 100.0;
+
+            // Define some point on the triangle
+            let point_on_triangle = triangle.0 + u * u_vec + v * v_vec;
+
+            // Define a ray which points at the triangle
+            let origin = tuple_to_point(&origin);
+            let ray = Ray::new(origin, point_on_triangle - origin);
+            let on_back_side = normal.dot(&(ray.origin - triangle.0)) <= 0.0;
+
+            // Compute
+            let intersects = ray.intersects_triangle(&triangle.0, &triangle.1, &triangle.2);
+            let uv_sum = intersects.1 + intersects.2;
+
+            // Either the intersection is in the back side (including the triangle-plane)
+            if on_back_side {
+                // Intersection must be INFINITY, u and v are undefined
+                intersects.0 == INFINITY
+            } else {
+                // Or it is on the front side
+                // Either the intersection is inside the triangle, which it should be
+                // for all u, v such that u+v <= 1.0
+                let intersection_inside = uv_sum >= 0.0 && uv_sum <= 1.0 && intersects.0 < INFINITY;
+
+                // Or the input data was close to the border
+                let close_to_border =
+                    u.abs() < EPSILON || (u - 1.0).abs() < EPSILON || v.abs() < EPSILON ||
+                    (v - 1.0).abs() < EPSILON || (u + v - 1.0).abs() < EPSILON;
+
+                if !(intersection_inside || close_to_border) {
```

Comment:
Performance is irrelevant in a test, but in case it fails it would be good to have some information on why it failed. The print is only executed if the test is about to fail.